### PR TITLE
add frontend for jax.numpy.fft.fft

### LINKED ISSUE #15832: fft

### PR DESCRIPTION
**Pull request by farhanoktvn**
Close #15832

### CODE REVIEW

**Comment by farhanoktvn** [comment]
The test failed for tensorflow backend at:
```
ivy/functional/backends/tensorflow/experimental/layers.py:479: in _fft_norm
    return x / tf.sqrt(n)
```

with error:
```
tensorflow.python.framework.errors_impl.InvalidArgumentError: Value for attr 'T' of int32 is not in the list of allowed values: bfloat16, half, float, double, complex64, complex128
```
```
NodeDef: {{node Sqrt}}; Op<name=Sqrt; signature=x:T -> y:T; attr=T:type,allowed=[DT_BFLOAT16, DT_HALF, DT_FLOAT, DT_DOUBLE, DT_COMPLEX64, DT_COMPLEX128]> [Op:Sqrt]
```

I tried adding `with_unsupported_dtypes` for `int32` on tensorflow backend, but it didn't resolve the issue. Do you have any suggestion on how to approach this problem?

File: ivy/functional/frontends/jax/numpy/fft.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# local
+import ivy
+from ivy.functional.frontends.jax.func_wrapper import to_ivy_arrays_and_back
+from ivy.func_wrapper import with_unsupported_dtypes
+
+
+@with_unsupported_dtypes({"1.24.3 and below": ("complex64",)}, "numpy")
```
Comment (hello-fri-end) [inline ivy/functional/frontends/jax/numpy/fft.py:7]:
```suggestion
@with_unsupported_dtypes({"1.24.3 and below": ("complex64",)}, "jax")
```

File: ivy_tests/test_ivy/test_frontends/test_jax/test_jax_numpy_fft.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+# local
+import ivy_tests.test_ivy.helpers as helpers
+from ivy_tests.test_ivy.helpers import handle_frontend_test
+
+
+# fft
+@handle_frontend_test(
+    fn_tree="jax.numpy.fft.fft",
+    dtype_values_axis=helpers.dtype_values_axis(
+        available_dtypes=helpers.get_dtypes("complex"),
```
Comment (hello-fri-end) [inline ivy_tests/test_ivy/test_frontends/test_jax/test_jax_numpy_fft.py:13]:
```suggestion
        available_dtypes=helpers.get_dtypes("valid"),
```
`valid` should be able to return all the data types supported by the function even if it's only complex.

**Comment by farhanoktvn** [comment]
I now have some nan/inf values from the test results:
```
 AssertionError:  the results from backend tensorflow and ground truth framework jax do not match
E            [ 5.92515092e+291-1.12355821e+307j -1.06856736e+307-3.47198581e+306j
E                        -inf            +infj  6.60410946e+306+9.08977685e+306j
E             1.06856736e+307-3.47198581e+306j]!=[-2.00000000e+000-1.12355821e+307j -1.06856736e+307-3.47198581e+306j
E            -6.60410946e+306+9.08977685e+306j  6.60410946e+306+9.08977685e+306j
E             1.06856736e+307-3.47198581e+306j] 
```
and
```
[-1.88708839e+307-1.63426649e+308j  9.43544196e+306-1.47083984e+308j
E             9.43544196e+306            -infj]!=[-1.88708839e+307-1.63426649e+308j  9.43544196e+306-1.47083984e+308j
E                         nan            -infj]
```

**Comment by farhanoktvn** [comment]
Also, I think there's an issue on `ivy/functional/backends/tensorflow/experimental/layers.py`.
```
def _fft_norm(
    x: Union[tf.Tensor, tf.Variable],
    dim: int,
    /,
    *,
    norm: str = "backward",
):
    n = tf.constant(x.shape[dim])
    if norm == "backward":
        return x
    elif norm == "ortho":
        return x / tf.sqrt(n)
    elif norm == "forward":
        return x / n
    else:
        raise ivy.utils.exceptions.IvyError(f"Unrecognized normalization mode {norm}")
```

When the norm is orthogonal, it returns _x_ divided by _tf.sqrt(n)_. However, _tf.sqrt_ does not accept _n_ as type int32. One possible way to solve this issue is to cast _n_, such that `n = tf.cast(n, x.dtype)`

**Review by hello-fri-end — CHANGES_REQUESTED** [review]

**Comment by hello-fri-end** [comment]
> Also, I think there's an issue on `ivy/functional/backends/tensorflow/experimental/layers.py`.
> 
> ```
> def _fft_norm(
>     x: Union[tf.Tensor, tf.Variable],
>     dim: int,
>     /,
>     *,
>     norm: str = "backward",
> ):
>     n = tf.constant(x.shape[dim])
>     if norm == "backward":
>         return x
>     elif norm == "ortho":
>         return x / tf.sqrt(n)
>     elif norm == "forward":
>         return x / n
>     else:
>         raise ivy.utils.exceptions.IvyError(f"Unrecognized normalization mode {norm}")
> ```
> 
> When the norm is orthogonal, it returns _x_ divided by _tf.sqrt(n)_. However, _tf.sqrt_ does not accept _n_ as type int32. One possible way to solve this issue is to cast _n_, such that `n = tf.cast(n, x.dtype)`

Hey @farhanoktvn ! Thanks for opening this PR. There are few problems `ivy.fft` which I'm currently fixing in https://github.com/unifyai/ivy/pull/16135. Will let you know once it's merged so that we can figure out the problems in the frontends :)

File: ivy/functional/frontends/jax/numpy/fft.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+            "complex64",
+            "complex128",
+        )
+    },
+    "jax",
+)
+@to_ivy_arrays_and_back
+def fft(a, n=None, axis=-1, norm=None):
+    res = ivy.fft(a, axis, norm=norm, n=n)
+    return ivy.astype(res, a.dtype)
```
Comment (hello-fri-end) [inline ivy/functional/frontends/jax/numpy/fft.py:19]:
```suggestion
    if norm == None:
       norm == 'backward
    return ivy.fft(a, axis, norm=norm, n=n)
```

File: ivy_tests/test_ivy/test_frontends/test_jax/test_jax_numpy_fft.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    fn_tree="jax.numpy.fft.fft",
+    dtype_values_axis=helpers.dtype_values_axis(
+        available_dtypes=helpers.get_dtypes("valid"),
+        num_arrays=1,
+        shape=(2,),
+        min_axis=-1,
+        force_int_axis=True,
+    ),
+    n=st.integers(min_value=2, max_value=10),
+    norm=st.sampled_from(["backward", "ortho", "forward"]),
```
Comment (hello-fri-end) [inline ivy_tests/test_ivy/test_frontends/test_jax/test_jax_numpy_fft.py:20]:
```suggestion
    norm=st.sampled_from(["backward", "ortho", "forward", None]),
```
We should also test for `norm=None`

File: ivy_tests/test_ivy/test_frontends/test_jax/test_jax_numpy_fft.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    test_with_out=st.just(False),
```
Comment (hello-fri-end) [inline ivy_tests/test_ivy/test_frontends/test_jax/test_jax_numpy_fft.py:21]:
```suggestion
    test_with_out=st.just(False),
```
No need for this :)

**Comment by hello-fri-end** [comment]
> > Also, I think there's an issue on `ivy/functional/backends/tensorflow/experimental/layers.py`.
> > ```
> > def _fft_norm(
> >     x: Union[tf.Tensor, tf.Variable],
> >     dim: int,
> >     /,
> >     *,
> >     norm: str = "backward",
> > ):
> >     n = tf.constant(x.shape[dim])
> >     if norm == "backward":
> >         return x
> >     elif norm == "ortho":
> >         return x / tf.sqrt(n)
> >     elif norm == "forward":
> >         return x / n
> >     else:
> >         raise ivy.utils.exceptions.IvyError(f"Unrecognized normalization mode {norm}")
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > When the norm is orthogonal, it returns _x_ divided by _tf.sqrt(n)_. However, _tf.sqrt_ does not accept _n_ as type int32. One possible way to solve this issue is to cast _n_, such that `n = tf.cast(n, x.dtype)`
> 
> Hey @farhanoktvn ! Thanks for opening this PR. There are few problems `ivy.fft` which I'm currently fixing in #16135. Will let you know once it's merged so that we can figure out the problems in the frontends :)

Hey @farhanoktvn ! The issues have been fixed. Once you update the code as per the changes I've suggested, you shouldn't get any errors. Feel free to request another review when it's done. Thanks :)

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Comment by farhanoktvn** [comment]
Hello, Bot. I'm still working on this PR.

**Comment by hello-fri-end** [comment]
> Hello, Bot. I'm still working on this PR.

Hey @farhanoktvn ! I have removed the stale label from the PR :)

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Comment by farhanoktvn** [comment]
I'm still working on this, sry it takes a while

**Comment by farhanoktvn** [comment]
Syncing my repo to the updated directory tree.
